### PR TITLE
[WINDOWING] Created main window structure.

### PR DESCRIPTION
**Pull request by sebawoz02**

### CODE REVIEW

File: src/windowing/window.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    glfwSetCursorPosCallback(glfw_window, cursor_position_callbacks);
+  }
+
+  Window::~Window()
+  {
+    glfwDestroyWindow(glfw_window);
+    glfwTerminate();
+  }
+
+  /* Public methods */
```
Comment (kociap) [inline src/windowing/window.cpp]: unnecessary comment

File: src/windowing/window.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  u8 Window::make_current_context() const
```
Comment (kociap) [inline src/windowing/window.cpp]: based on the name of the function, what does this function do? can't tell

File: src/windowing/window.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  {
+    for(auto& rect: movable_rectangles) {
+      rect.render();
+    }
+    if(custom_render != nullptr) {
+      custom_render();
+    }
+  }
+
+  /* Static methods */
```
Comment (kociap) [inline src/windowing/window.cpp]: unnecessary comment

File: src/windowing/window.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+  void Window::cursor_position_callbacks(GLFWwindow* win, double xpos,
+                                         double ypos)
+  {
+    auto* instance = static_cast<Window*>(glfwGetWindowUserPointer(win));
+
+    instance->cursor_handler(xpos, ypos);
+  }
+
+  /* Private methods */
```
Comment (kociap) [inline src/windowing/window.cpp]: unnecessary comment

File: src/windowing/window.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+      Vec2 offset;
+      offset.x = x - last_mouse_position.x;
+      offset.y = y - last_mouse_position.y;
+
+      currently_moved_rectangle->move(offset);
+      last_mouse_position.x = x;
+      last_mouse_position.y = y;
+    }
+  }
+} // namespace nebula
```
Comment (kociap) [inline src/windowing/window.cpp]: missing newline

File: src/windowing/window.hpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+#include <GLFW/glfw3.h>
+#include <types.hpp>
+#include <ui/movable_rect.hpp>
+
+namespace nebula {
+  struct Window {
+  public:
+    /* Variables */
+    GLFWwindow* glfw_window; // GLFW window instance
```
Comment (kociap) [inline src/windowing/window.hpp]: I believe these should not be public

File: src/windowing/window.hpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+     * ( Other then MR buttons and created by them rectangles. )
+     */
+    explicit Window(void (*custom_render_func)());
+
+    /**
+     * Destroys window and terminates glfw.
+     */
+    ~Window();
+
+    /* Methods */
```
Comment (kociap) [inline src/windowing/window.hpp]: unnecessary comment

File: src/windowing/window.hpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    /* Private Variables */
+
+    Vec2 last_mouse_position; // last mouse position
+    List<Movable_Rect> movable_rectangles; // All created rectangles/gates
+    Movable_Rect* currently_moved_rectangle; // Currently moved rect/gate
+
+    /* Private Methods */
+
+    /**
+     * Keyboard input handler.
```
Comment (kociap) [inline src/windowing/window.hpp]: this comment does not introduce any new information, hence it's redundant and unnecessary

File: src/windowing/window.hpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * Adds new rectangle. Function used by MR buttons.
+     * @param render_function - render function inherited from MR button.
+     * @param rectangle_dimensions - dimensions inherited from MR button.
+     */
+    void add_rectangle(void (*render_function)(Node_Rect),
+                       Vec2 rectangle_dimensions);
+
+    /** Static methods
+     * These are template callback functions redirecting to non-static handlers.
```
Comment (kociap) [inline src/windowing/window.hpp]: purpose of these? why can't those be hidden in the source file as static linkage non-member functions?

File: src/windowing/window.cpp [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        }
+      } else if(action == GLFW_RELEASE) {
+        currently_moved_rectangle = nullptr;
+      }
+    }
+  }
+
+  void Window::cursor_handler(double xpos, double ypos)
+  {
+    if(currently_moved_rectangle != nullptr) {
```
Comment (kociap) [inline src/windowing/window.cpp]: `if(currently_moved_rectangle == nullptr) { return; }`
reduces nesting

**Review by kociap — CHANGES_REQUESTED** [review]

File: src/windowing/window.cpp (hunk 2 above)
Comment (sebawoz02) [reply src/windowing/window.cpp]: It's unnecessary tbh if we are using only one window. I'm moving this to the constructor.

**Review by kociap — APPROVED** [review]

**Pull request merged**